[WIP] Custom nginx server setup

nixterrimus opened the pull request:
After launching this proxy on heroku we had two problems:
- SSL was not being proxied properly
- Sailthru uses ports other than 80 and 443 to handle unsubscribes

Heroku, while wonderful for hosting things that live on port 80 and 443 does not handle servers that need to open (literally) 1000 ports.  The issues with SSL appear like they may have been related to the way heroku handles their routing.  This PR addresses these issues by moving this off of heroku and onto AWS.
## Introducing Packer

[Packer](http://packer.io) is a tool for building machine images.  Here we use to it to build an Amazon AMI.

To use this PR you'll first need to install packer (`brew install packer`)

After that you'll need to place the SSL certs into the `certs/` folder.

Finally run `packer packer.json`

nixterrimus committed:
Don’t reuse sailthru’s SSL cert on proxy

nixterrimus committed:
Proxy the SSL config of the app association host

nixterrimus committed:
Route request for root to the app association host

nixterrimus committed:
Listen with SSL

nixterrimus committed:
Turn off all SSL proxy reuse

nixterrimus committed:
Remove SSL

nixterrimus committed:
Bind to many ports for unsubscribing

nixterrimus committed:
Remove port settings, the server will set these things now

nixterrimus committed:
SSL Configuration

nixterrimus committed:
Add packer configuration for setting up nginx on Amazon

nixterrimus committed:
Add folder for certs and ignore certs from git

nixterrimus committed:
Copy SSL files into target machine

nixterrimus committed:
Compile nginx con and copy it to the target machine

nixterrimus committed:
Outline steps for completing server setup

dannygreg commented on deploy/server-setup.sh:
```diff
@@ -0,0 +1,20 @@
+sudo apt-get install nginx -y
+
+echo "packer: nginx - installing as a service"
+sudo update-rc.d nginx defaults
```
We're possibly better off using Upstart rather than `init.d` here. I don't know what system we will be installing it on but Upstart has replaced `init.d` in modern Ubuntu distros. In fact, old init scripts are run through a compatibility layer in Upstart, which can sometimes be a bit dodgy.

nixterrimus committed:
Ignore crt files

nixterrimus committed:
A working but incomplete deploy- this generates an API but doesn’t serve SSL properly

nixterrimus committed:
Only serve SSL on port 443, build complete nginx config

nixterrimus committed:
Copy nginx config the proper place

nixterrimus committed:
Use a newer version of ubuntu as the base

nixterrimus committed:
Ignore pem files

nixterrimus committed:
Handle unsubscribes via link rewriting

Rather than proxying the request, simply send the user directly to sailthru’s servers and let them handle the port mappings.

nixterrimus committed:
Rename amazon AMI
